File: net/pdfjet/Arc.cs
/*
 * Arc.cs
 *
 * Copyright (c) 2026 PDFjet Software
 * Licensed under the MIT License. See LICENSE file in the project root.
 */
using System;

namespace PDFjet.NET {
/// <summary>
/// Used to create arc objects.
/// </summary>
public class Arc : IDrawable {
    private float cx;
    private float cy;
    private float rx;
    private float ry;
    private float startAngle;
    private float sweepDegrees;
    private float rotateDegrees;

    private float[] fillColor;
    private float[] strokeColor = new float[] {0f, 0f, 0f};   // Black color
    private float strokeWidth = 0f;
    private String strokeDashPattern = "[] 0";

    private String language = null;
    private String actualText = Single.space;
    private String altDescription = Single.space;

    private Line line;

    /// <summary>
    /// The default constructor.
    /// </summary>
    public Arc() {
    }

    /// <summary>Sets the center of this arc.</summary>
    public Arc SetLocation(float cx, float cy) {
        SetCenterXY(cx, cy);
        return this;
    }

    IDrawable IDrawable.SetLocation(float cx, float cy) {
        return SetLocation(cx, cy);
    }

    /// <summary>Starts this arc at the end point of the specified line.</summary>
    public Arc SetStartPointToEndOf(Line line) {
        this.line = line;
        return this;
    }

    /// <summary>Sets the center of this arc.</summary>
    public Arc SetCenterXY(float cx, float cy) {
        this.cx = cx;
        this.cy = cy;
        return this;
    }

    /// <summary>Sets the horizontal radius of this arc.</summary>
    public Arc SetRadiusX(float rx) {
        this.rx = rx;
        return this;
    }

    /// <summary>Sets the vertical radius of this arc.</summary>
    public Arc SetRadiusY(float ry) {
        this.ry = ry;
        return this;
    }

    /// <summary>Sets both radii of this arc to the same value, making it circular.</summary>
    public Arc SetRadius(float r) {
        this.rx = r;
        this.ry = r;
        return this;
    }

    /// <summary>Sets the angle in degrees where this arc starts.</summary>
    public Arc SetStartAngle(float angle) {
        this.startAngle = angle;
        return this;
    }

    /// <summary>Sets how many degrees this arc sweeps clockwise from its start angle.</summary>
    public Arc SetSweepDegreesCW(float sweepDegrees) {
        this.sweepDegrees = sweepDegrees;
        return this;
    }

    /// <summary>Sets how many degrees this arc sweeps counterclockwise from its start angle.</summary>
    public Arc SetSweepDegreesCCW(float sweepDegrees) {
        this.sweepDegrees = -sweepDegrees;
        return this;
    }

    /// <summary>
    /// The line dash pattern controls the pattern of dashes and gaps used to stroke paths.
    /// It is specified by a dash array and a dash phase.
    /// The elements of the dash array are positive numbers that specify the lengths of
    /// alternating dashes and gaps.
    /// The dash phase specifies the distance into the dash pattern at which to start the dash.
    /// The elements of both the dash array and the dash phase are expressed in user space units.
    /// <code>
    /// Examples of line dash patterns:
    ///
    ///     "[Array] Phase"     Appearance          Description
    ///     _______________     _________________   ____________________________________
    ///     "[] 0"              -----------------   Solid line
    ///     "[3] 0"             ---   ---   ---     3 units on, 3 units off, ...
    ///     "[2] 1"             -  --  --  --  --   1 on, 2 off, 2 on, 2 off, ...
    ///     "[2 1] 0"           -- -- -- -- -- --   2 on, 1 off, 2 on, 1 off, ...
    ///     "[3 5] 6"             ---     ---       2 off, 3 on, 5 off, 3 on, 5 off, ...
    ///     "[2 3] 11"          -   --   --   --    1 on, 3 off, 2 on, 3 off, 2 on, ...
    /// </code>
    /// </summary>
    /// <param name="strokeDashPattern">the stroke dash pattern.</param>
    /// <returns>this Arc object.</returns>
    public Arc SetStrokeDashPattern(String strokeDashPattern) {
        this.strokeDashPattern = strokeDashPattern;
        return this;
    }

    /// <summary>
    /// Sets the width of this line.
    /// </summary>
    /// <param name="width">the width.</param>
    /// <returns>this Arc object.</returns>
    public Arc SetStrokeWidth(double width) {
        this.strokeWidth = (float) width;
        return this;
    }

    /// <summary>
    /// Sets the width of this line.
    /// </summary>
    /// <param name="width">the width.</param>
    /// <returns>this Arc object.</returns>
    public Arc SetStrokeWidth(float width) {
        this.strokeWidth = width;
        return this;
    }

    /// <summary>
    /// Sets the color for this line.
    /// </summary>
    /// <param name="color">the color specified as an integer.</param>
    /// <returns>this Arc object.</returns>
    public Arc SetStrokeColor(int color) {
        float r = ((color >> 16) & 0xff)/255f;
        float g = ((color >>  8) & 0xff)/255f;
        float b = ((color)       & 0xff)/255f;
        this.SetStrokeColor(r, g, b);
        return this;
    }

    /// <summary>Sets the stroke color from red, green and blue values between 0.0 and 1.0.</summary>
    public Arc SetStrokeColor(float r, float g, float b) {
        this.strokeColor = new float[] {r, g, b};
        return this;
    }

    /// <summary>Sets the stroke color from an array of red, green and blue values.</summary>
    public Arc SetStrokeColor(float[] rgbColor) {
        this.strokeColor = rgbColor;
        return this;
    }

    /// <summary>Sets the fill color as a 0xRRGGBB value, for example Color.blue.</summary>
    public Arc SetFillColor(int color) {
        float r = ((color >> 16) & 0xff)/255f;
        float g = ((color >>  8) & 0xff)/255f;
        float b = ((color)       & 0xff)/255f;
        this.SetFillColor(r, g, b);
        return this;
    }

    /// <summary>Sets the fill color from red, green and blue values between 0.0 and 1.0.</summary>
    public Arc SetFillColor(float r, float g, float b) {
        this.fillColor = new float[] {r, g, b};
        return this;
    }

    /// <summary>Sets the fill color from an array of red, green and blue values.</summary>
    public Arc SetFillColor(float[] rgbColor) {
        this.fillColor = rgbColor;
        return this;
    }

    /// <summary>Rotates this arc clockwise by the specified degrees.</summary>
    public Arc SetRotateDegreesCW(float degrees) {
        this.rotateDegrees = -degrees;
        return this;
    }

    /// <summary>Rotates this arc clockwise by the specified degrees.</summary>
    public Arc SetRotateDegreesCW(double degrees) {
        this.rotateDegrees = (float) -degrees;
        return this;
    }

    /// <summary>Rotates this arc counterclockwise by the specified degrees.</summary>
    public Arc SetRotateDegreesCCW(float degrees) {
        this.rotateDegrees = degrees;
        return this;
    }

    /// <summary>Rotates this arc counterclockwise by the specified degrees.</summary>
    public Arc SetRotateDegreesCCW(double degrees) {
        this.rotateDegrees = (float) degrees;
        return this;
    }

    /// <summary>
    /// Sets the alternate description of this line.
    /// </summary>
    /// <param name="altDescription">the alternate description of the line.</param>
    /// <returns>this Arc.</returns>
    public Arc SetAltDescription(String altDescription) {
        this.altDescription = altDescription;
        return this;
    }

    /// <summary>
    /// Sets the actual text for this line.
    /// </summary>
    /// <param name="actualText">the actual text for the line.</param>
    /// <returns>this Arc.</returns>
    public Arc SetActualText(String actualText) {
        this.actualText = actualText;
        return this;
    }

    /// <summary>
    /// Scales this line by the specified factor.
    /// </summary>
    /// <param name="factor">the factor used to scale the line.</param>
    /// <returns>this Arc object.</returns>
    public Arc SetScaleFactor(double factor) {
        return SetScaleFactor((float) factor);
    }

    /// <summary>
    /// Scales this line by the specified factor.
    /// </summary>
    /// <param name="factor">the factor used to scale the line.</param>
    /// <returns>this Arc object.</returns>
    public Arc SetScaleFactor(float factor) {
        this.rx *= factor;
        this.ry *= factor;
        return this;
    }

    /// <summary>
    /// Draws this line on the specified page.
    /// </summary>
    /// <param name="page">the page to draw on.</param>
    /// <returns>x and y coordinates of the bottom right corner of this component.</returns>
    public float[] DrawOn(Page page) {
        // If a start point was set, calculate center so arc begins there
        if (line != null) {
            float dx = line.x2 - line.x1;
            float dy = line.y2 - line.y1;
            // Normalize and rotate 90° (clockwise perpendicular)
            float invLength = 1f / MathF.Sqrt(dx*dx + dy*dy);
            float nx = -dy * invLength;
            float ny = dx * invLength;
            // Adjust direction based on sweep
            float sign = sweepDegrees > 0f ? 1f : -1f;
            cx = line.x2 + nx * rx * sign;
            cy = line.y2 + ny * ry * sign;
            startAngle = MathF.Atan2(line.y2 - cy, line.x2 - cx) * (180f / MathF.PI);
        }

        page.AddBMC(StructElem.P, language, actualText, altDescription);

        page.SaveGraphicsState();

        float centerX = cx;
        float centerY = page.height - cy;
        page.RotateAroundCenter(centerX, centerY, rotateDegrees);
        page.DrawArc(
                cx,
                cy,
                rx,
                ry,
                startAngle,
                sweepDegrees);
        if (strokeColor != null && strokeDashPattern != null) {
            page.SetStrokeDashPattern(strokeDashPattern);
        }
        if (fillColor != null && strokeColor != null) {
            page.SetBrushColor(fillColor);
            page.SetPenWidth(strokeWidth);
            page.SetPenColor(strokeColor);
            page.Append("B\n");
        } else if (fillColor != null && strokeColor == null) {
            page.SetBrushColor(fillColor);
            page.Append("f\n");
        } else if (fillColor == null && strokeColor != null) {
            page.SetPenWidth(strokeWidth);
            page.SetPenColor(strokeColor);
            page.Append("S\n");
        } else {    // Both brushColor == null and penColor == null
            page.SetPenWidth(0f);
            page.SetPenColor(Color.black);
            page.Append("S\n");
        }

        page.RestoreGraphicsState();

        page.AddEMC();
        return new float[] {cx + rx, cy + ry};
    }
}   // End of Arc.cs
}   // End of namespace PDFjet.NET

File: Sources/PDFjet/Arc.swift
/**
 *  Arc.swift
 *
 *  Copyright (c) 2026 PDFjet Software
 *  Licensed under the MIT License. See LICENSE file in the project root.
 */
import Foundation

///
/// Used to create arc objects.
///
public class Arc : Drawable {
    private var cx: Float = 0.0
    private var cy: Float = 0.0
    private var rx: Float = 0.0
    private var ry: Float = 0.0
    private var startAngle: Float = 0.0
    private var sweepDegrees: Float = 0.0
    private var rotateDegrees: Float = 0.0

    private var fillColor: [Float]?
    private var strokeColor: [Float]? = [0.0, 0.0, 0.0]     // Black color
    private var strokeWidth: Float = 0.0
    private var strokeDashPattern: String? = "[] 0"

    private var language: String?
    private var actualText: String = Single.space
    private var altDescription: String = Single.space

    private var line: Line?

    ///
    /// The default constructor.
    ///
    public init() {
    }

    /// Sets the center of this arc.
    @discardableResult
    public func setLocation(_ cx: Float, _ cy: Float) -> Self {
        _ = setCenterXY(cx, cy)
        return self
    }

    /// Starts this arc at the end point of the specified line.
    @discardableResult
    public func setStartPointToEndOf(_ line: Line) -> Arc {
        self.line = line
        return self
    }

    /// Sets the center of this arc.
    @discardableResult
    public func setCenterXY(_ cx: Float, _ cy: Float) -> Arc {
        self.cx = cx
        self.cy = cy
        return self
    }

    /// Sets the horizontal radius of this arc.
    @discardableResult
    public func setRadiusX(_ rx: Float) -> Arc {
        self.rx = rx
        return self
    }

    /// Sets the vertical radius of this arc.
    @discardableResult
    public func setRadiusY(_ ry: Float) -> Arc {
        self.ry = ry
        return self
    }

    /// Sets both radii of this arc to the same value, making it circular.
    @discardableResult
    public func setRadius(_ r: Float) -> Arc {
        self.rx = r
        self.ry = r
        return self
    }

    /// Sets the angle in degrees where this arc starts.
    @discardableResult
    public func setStartAngle(_ angle: Float) -> Arc {
        self.startAngle = angle
        return self
    }

    /// Sets how many degrees this arc sweeps clockwise from its start angle.
    @discardableResult
    public func setSweepDegreesCW(_ sweepDegrees: Float) -> Arc {
        self.sweepDegrees = sweepDegrees
        return self
    }

    /// Sets how many degrees this arc sweeps counterclockwise from its start angle.
    @discardableResult
    public func setSweepDegreesCCW(_ sweepDegrees: Float) -> Arc {
        self.sweepDegrees = -sweepDegrees
        return self
    }

    ///
    /// The line dash pattern controls the pattern of dashes and gaps used to stroke paths.
    /// It is specified by a dash array and a dash phase.
    /// The elements of the dash array are positive numbers that specify the lengths of
    /// alternating dashes and gaps.
    /// The dash phase specifies the distance into the dash pattern at which to start the dash.
    /// The elements of both the dash array and the dash phase are expressed in user space units.
    ///
    ///     Examples of line dash patterns:
    ///
    ///         "[Array] Phase"     Appearance          Description
    ///         _______________     _________________   ____________________________________
    ///         "[] 0"              -----------------   Solid line
    ///         "[3] 0"             ---   ---   ---     3 units on, 3 units off, ...
    ///         "[2] 1"             -  --  --  --  --   1 on, 2 off, 2 on, 2 off, ...
    ///         "[2 1] 0"           -- -- -- -- -- --   2 on, 1 off, 2 on, 1 off, ...
    ///         "[3 5] 6"             ---     ---       2 off, 3 on, 5 off, 3 on, 5 off, ...
    ///         "[2 3] 11"          -   --   --   --    1 on, 3 off, 2 on, 3 off, 2 on, ...
    ///
    /// - Parameter strokeDashPattern: the line dash pattern.
    /// - Returns: this Arc object.
    ///
    @discardableResult
    public func setStrokeDashPattern(_ strokeDashPattern: String) -> Arc {
        self.strokeDashPattern = strokeDashPattern
        return self
    }

    ///
    /// Sets the width of this line.
    ///
    /// - Parameter strokeWidth: the width.
    /// - Returns: this Arc object.
    ///
    @discardableResult
    public func setStrokeWidth(_ strokeWidth: Float) -> Arc {
        self.strokeWidth = strokeWidth
        return self
    }

    ///
    /// Sets the color for this line.
    ///
    /// - Parameter color: the color specified as an integer.
    /// - Returns: this Arc object.
    ///
    @discardableResult
    public func setStrokeColor(_ color: Int32) -> Arc {
        let r = Float((color >> 16) & 0xff)/255.0
        let g = Float((color >>  8) & 0xff)/255.0
        let b = Float((color)       & 0xff)/255.0
        return setStrokeColor(r, g, b)
    }

    /// Sets the stroke color from red, green and blue values between 0.0 and 1.0.
    @discardableResult
    public func setStrokeColor(_ r: Float, _ g: Float, _ b: Float) -> Arc {
        self.strokeColor = [r, g, b]
        return self
    }

    /// Sets the stroke color from an array of red, green and blue values.
    @discardableResult
    public func setStrokeColor(_ rgbColor: [Float]) -> Arc {
        self.strokeColor = rgbColor
        return self
    }

    /// Sets the fill color as a 0xRRGGBB value, for example Color.blue.
    @discardableResult
    public func setFillColor(_ color: Int32) -> Arc {
        let r = Float((color >> 16) & 0xff)/255.0
        let g = Float((color >>  8) & 0xff)/255.0
        let b = Float((color)       & 0xff)/255.0
        return setFillColor(r, g, b)
    }

    /// Sets the fill color from red, green and blue values between 0.0 and 1.0.
    @discardableResult
    public func setFillColor(_ r: Float, _ g: Float, _ b: Float) -> Arc {
        self.fillColor = [r, g, b]
        return self
    }

    /// Sets the fill color from an array of red, green and blue values.
    @discardableResult
    public func setFillColor(_ rgbColor: [Float]) -> Arc {
        self.fillColor = rgbColor
        return self
    }

    /// Rotates this arc clockwise by the specified degrees.
    @discardableResult
    public func setRotateDegreesCW(_ degrees: Float) -> Arc {
        self.rotateDegrees = -degrees
        return self
    }

    /// Rotates this arc counterclockwise by the specified degrees.
    @discardableResult
    public func setRotateDegreesCCW(_ degrees: Float) -> Arc {
        self.rotateDegrees = degrees
        return self
    }

    ///
    /// Sets the alternate description of this line.
    ///
    /// - Parameter altDescription: the alternate description of the line.
    /// - Returns: this Arc.
    ///
    @discardableResult
    public func setAltDescription(_ altDescription: String) -> Arc {
        self.altDescription = altDescription
        return self
    }

    ///
    /// Sets the actual text for this line.
    ///
    /// - Parameter actualText: the actual text for the line.
    /// - Returns: this Arc.
    ///
    @discardableResult
    public func setActualText(_ actualText: String) -> Arc {
        self.actualText = actualText
        return self
    }

    ///
    /// Scales this line by the specified factor.
    ///
    /// - Parameter factor: the factor used to scale the line.
    /// - Returns: this Arc object.
    ///
    @discardableResult
    public func setScaleFactor(_ factor: Float) -> Arc {
        self.rx *= factor
        self.ry *= factor
        return self
    }

    ///
    /// Draws this line on the specified page.
    ///
    /// - Parameter page: the page to draw on.
    /// - Returns: x and y coordinates of the bottom right corner of this component.
    ///
    @discardableResult
    public func drawOn(_ page: Page?) -> [Float] {
        // If a start point was set, calculate center so arc begins there
        if line != nil {
            let p1 = line!.getStartPoint()
            let p2 = line!.getEndPoint()
            let dx = p2.x - p1.x
            let dy = p2.y - p1.y
            // Normalize and rotate 90 degrees (clockwise perpendicular)
            let invLength = 1.0 / sqrt(dx*dx + dy*dy)
            let nx = -dy * invLength
            let ny = dx * invLength
            // Adjust direction based on sweep
            let sign: Float = sweepDegrees > 0.0 ? 1.0 : -1.0
            cx = p2.x + nx * rx * sign
            cy = p2.y + ny * ry * sign
            startAngle = atan2(p2.y - cy, p2.x - cx) * (180.0 / Float.pi)
        }

        page!.addBMC(StructElem.P, language, actualText, altDescription)
        page!.saveGraphicsState()
        let centerX = cx
        let centerY = page!.height - cy
        page!.rotateAroundCenter(centerX, centerY, rotateDegrees)
        _ = page!.drawArc(cx, cy, rx, ry, startAngle, sweepDegrees)
        if strokeColor != nil && strokeDashPattern != nil {
            page!.setStrokeDashPattern(strokeDashPattern!)
        }
        if fillColor != nil && strokeColor != nil {
            page!.setBrushColor(fillColor!)
            page!.setPenWidth(strokeWidth)
            page!.setPenColor(strokeColor!)
            page!.append("B\n")
        } else if fillColor != nil && strokeColor == nil {
            page!.setBrushColor(fillColor!)
            page!.append("f\n")
        } else if fillColor == nil && strokeColor != nil {
            page!.setPenWidth(strokeWidth)
            page!.setPenColor(strokeColor!)
            page!.append("S\n")
        } else {    // Both brushColor == nil and penColor == nil
            page!.setPenWidth(0.0)
            page!.setPenColor(Color.black)
            page!.append("S\n")
        }
        page!.restoreGraphicsState()
        page!.addEMC()
        return [cx + rx, cy + ry]
    }
}   // End of Arc.swift
